quil-rs/src/instruction/calibration.rs

```diff
@@ -0,0 +1,92 @@
+use std::fmt;
+
+use crate::{
+    instruction::{
+        format_instructions, format_qubits, get_expression_parameter_string, Expression,
+        GateModifier, Instruction, Qubit,
+    },
+    validation::identifier::{validate_identifier, IdentifierValidationError},
+};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Calibration {
+    pub instructions: Vec<Instruction>,
+    pub modifiers: Vec<GateModifier>,
+    pub name: String,
+    pub parameters: Vec<Expression>,
+    pub qubits: Vec<Qubit>,
+}
+
+impl Calibration {
+    /// Builds a new calibration definition.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the given name isn't a valid Quil identifier.
+    pub fn new(
+        name: &str,
+        parameters: Vec<Expression>,
+        qubits: Vec<Qubit>,
+        instructions: Vec<Instruction>,
+        modifiers: Vec<GateModifier>,
+    ) -> Result<Self, IdentifierValidationError> {
+        validate_identifier(name)?;
+        Ok(Self {
+            instructions,
+            modifiers,
+            name: name.to_string(),
+            parameters,
+            qubits,
+        })
+    }
+}
+
+impl fmt::Display for Calibration {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let parameter_str = get_expression_parameter_string(&self.parameters);
+        write!(
+            f,
+            "DEFCAL {}{} {}:",
+            self.name,
+            parameter_str,
+            format_qubits(&self.qubits)
+        )?;
+        for instruction in &self.instructions {
+            write!(f, "\n\t{instruction}")?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MeasureCalibrationDefinition {
+    pub qubit: Option<Qubit>,
+    pub parameter: String,
+    pub instructions: Vec<Instruction>,
+}
+
+impl MeasureCalibrationDefinition {
+    pub fn new(qubit: Option<Qubit>, parameter: String, instructions: Vec<Instruction>) -> Self {
+        Self {
+            qubit,
+            parameter,
+            instructions,
+        }
+    }
+}
```

Comment:
Maybe not necessary, but it makes the API more consistent if all the instruction types have new methods. Making the members of the instructions private is up for consideration as well since manual manipulation of the fields could easily malform an instruction. That would be a big change, but having and using constructors allows us to make it much easier in the future.